Allow manual selection of units on results entry

RML-IAEA opened the pull request:
+ A list of units can be defined for an analysis service
+ The user can select the units when submitting data.

## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/

## Current behavior before PR

## Desired behavior after PR is merged

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

RML-IAEA committed:
Unit selection

+ A list of units can be defined for an analysis service
+ The user can select the units when submitting data.

RML-IAEA committed:
Copy of line error _on_method_change

RML-IAEA committed:
Update view.py (default listed first, text-nowrap unit)

update the code so that the default value is always listed first.
update the code so that the units rendered beside the result do not wrap.

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -1543,6 +1601,16 @@ def is_instrument_required(self, analysis):
+        # Always return true if the analysis has unitchoices
+        analysis = self.get_object(analysis)
+        if analysis.getUnitChoices():
+            return True
```
Please add a `return False` for when the condition is not met

RML-IAEA replied:
Implemented

xispa commented on src/bika/lims/content/abstractbaseanalysis.py:
```diff
@@ -753,6 +791,13 @@ def getUnit(self):
+        '''Returns the UnitChoices
+        :returns: List of UnitChoices objects
+        '''
+        return self.Schema().getField("UnitChoices").get(self) or ""
```
Note Plone automatically adds accessors for fields and we only add our own when we do want to override the default behavior. I think this is not the case, please remove the accessor, and also note that `RecordsField` always return empty list as default when there is no value set.

RML-IAEA replied:
Thank you, custom accessors removed.

xispa commented on src/bika/lims/validators.py:
```diff
@@ -1507,3 +1507,44 @@ def validate_record(self, record):
+        unit_choice = record.get("unitchoice")
+        # Check that the unit choice is a string
+        if not unit_choice:
+                    return _("Validation failed: '{}' is not a string").format(
```
No valid indentation here, 12 whitespace characters required

RML-IAEA replied:
Thank you, validator removed.

xispa commented on src/bika/lims/validators.py:
```diff
@@ -1507,3 +1507,44 @@ def validate_record(self, record):
 
 validation.register(ServiceConditionsValidator())
+
+class ServiceUnitChoicesValidator(object):
```
This validator is not required. Since you already defined the `unitchoice` type to be a `string` in the field definition (`subfield_types`  key), the records-specific validator that comes by default will apply already.

RML-IAEA replied:
Thank you, validator removed.

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -135,6 +135,12 @@ def __init__(self, context, request, **kwargs):
                 "title": _("Calculation"),
                 "sortable": False,
                 "toggle": False}),
+            ("UnitChoices", {
```
Please add the `UnitChoices` field after `Result`, so the dropdown list for the selection of unit is displayed 
after the `Result` column, that imho looks way better:

![Captura de 2022-12-20 13-09-23](https://user-images.githubusercontent.com/832627/208664086-208ab98b-5162-4df8-968b-53948489b219.png)


instead of

![Captura de 2022-12-20 13-06-24](https://user-images.githubusercontent.com/832627/208663589-ff902d83-25a8-4e04-b53c-cd0c64f56f01.png)

RML-IAEA replied:
Changed to "Unit: and placed after uncertainty. Note that I changed the location of the uncertainty variable to be beside the result.

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -1045,6 +1054,44 @@ def _folder_item_calculation(self, analysis_brain, item):
+                    choices=["{}:{}".format(x['unitchoice'],x['unitchoice'].lstrip()) for x in unit_choice_fields]
+                    # Add default unit to list of choices
+                    # Space in {}: {} ensures that the default unit is always on top of the selection.
+                    choices.append("{}: {}".format(unit,unit))
```
I don't get all this workaround with leading white spaces to make the default unit to be the first in the list. Why to not simple insert the default unit at position 0 from the resulting displaylist `dl`?

RML-IAEA replied:
Thank you. I think I had sortable set to True when this was implemented. I have since set it to False and it is working fine.

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -1045,6 +1054,44 @@ def _folder_item_calculation(self, analysis_brain, item):
+            # Hide unit_choices if user cannot view results
+            return
+        is_editable = self.is_analysis_edition_allowed(analysis_brain)
+        if is_editable:
```
Note that we quite often do the other way round to prevent too much indentations. E.g.:

```python
if not self.is_analysis_edition_allowed(analysis_brain):
    return 

# Set unit_choices. The default unit is also included. 
analysis_obj = self.get_object(analysis_brain)
unit_choice_fields = analysis_obj.getUnitChoices() or list()
if not unit_choice_fields:
    return

# Get the keys for the dictionary
unit_choice_fields_key=set().union(*(d.keys() for d in unit_choice_fields))
[...]
```

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -135,6 +135,12 @@ def __init__(self, context, request, **kwargs):
                 "sortable": False,
                 "toggle": False}),
+            ("UnitChoices", {
+                "title": _("Unit Selection"),
```
Better "Unit" than "Unit Selection". Will save space, and the user will see that she/he has to select from a list anyways.

RML-IAEA replied:
Thank you, I have updated this.

xispa reviewed (changes requested):
Thanks for your efforts!

Seems the unit is not store properly

1. Type result:

![Captura de 2022-12-20 13-41-00](https://user-images.githubusercontent.com/832627/208670033-3a1de9a0-3db5-4b5d-a2b6-08ffb84ffb4a.png)

2. Select a unit from the selection list:

![Captura de 2022-12-20 13-41-15](https://user-images.githubusercontent.com/832627/208670114-0a65ac94-dd7c-4f1c-b945-5000e12367b1.png)

3. On save, the unit next to the result changes to default, different from the selected one:

![Captura de 2022-12-20 13-41-28](https://user-images.githubusercontent.com/832627/208670131-cb2856b3-b92b-4f38-b1b7-9bcb6f6278fd.png)

4. On submit, the unit has not changed, although the selected one was different from default:

![Captura de 2022-12-20 13-41-39](https://user-images.githubusercontent.com/832627/208670144-1516d2ae-ea18-41db-ba21-ef46e32fda33.png)

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -1065,7 +1112,18 @@ def _folder_item_method(self, analysis_brain, item):
                 item["Method"] = api.get_title(method)
                 item["replace"]["Method"] = get_link_for(method, tabindex="-1")
 
+    def _on_unit_choice_change(self, uid=None, value=None, item=None, **kw):
```
You don't need this if you do it differently. Hint: You don't need a "UnitChoices" column, but a "Unit" column that becomes editable (with a choices list) when the analysis have multiple units set. Otherwise, the column "Unit" is not rendered.

RML-IAEA replied:
Removed UnitChoice column. Unit it column is only rendered when needed.

xispa replied:
I think you can remove this function `_on_unit_choice_change`, cause `on_unit_change` is used instead

RML-IAEA replied:
Thank you, removed.

RML-IAEA committed:
Implemented Corrections

Implemented corrections from https://github.com/senaite/senaite.core/pull/2201#pullrequestreview-1224345271

RML-IAEA commented:
Thank you @xispa  for all your comments. I have implemented your comments (https://github.com/senaite/senaite.core/pull/2201/commits/ac745f925d630f9191d7cc7aad5efe13c5e7b775). 
A default unit can be provided here:
![image](https://user-images.githubusercontent.com/95472242/212373274-afbd61cc-9345-4bc6-b03b-f40b546dd8b4.png)

A list of units for selection can be provided here:
![image](https://user-images.githubusercontent.com/95472242/212373320-1ce79b08-88c2-4c60-ae20-d77ecba246b3.png)

The unit column will only render if a selection of units is provided. 
![image](https://user-images.githubusercontent.com/95472242/212373438-216595e7-6833-406a-9763-09534acaa027.png)

I have also updated the code so that the units render after the uncertainty.
![image](https://user-images.githubusercontent.com/95472242/212373548-f4d156ef-90e2-4dbc-888e-aee2a26e66f0.png)

and for the selection list;
![image](https://user-images.githubusercontent.com/95472242/212373727-a6a114bb-dff4-4f0e-9f8f-2ae68eca7343.png)


The units change is now stored properly:

![image](https://user-images.githubusercontent.com/95472242/212373840-e2862b5e-45de-4893-a055-d499f26f6aae.png)

On submit:
![image](https://user-images.githubusercontent.com/95472242/212373922-4a3e136e-c34f-4c92-b8f3-7ed85f6967b2.png)



Improvement:
I wanted to include the unitchoice object with the unit object. I tried to implement this by updating the unit object to be a RecordsField. I was unable to set the selected value properly. I notice that for UIDReferenceField this works well. Do you have any suggestions on how the unitchoice object could be merged with the unit object?

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into unitSelection

xispa commented on src/bika/lims/browser/analyses/view.py:
```diff
@@ -1122,6 +1190,19 @@ def _folder_item_instrument(self, analysis_brain, item):
+
+    def _on_unit_change(self, uid=None, value=None, item=None, **kw):
+        """ updates the rendered unit on selection of unit. 
+        """
```
Note that value and item are declared as `None` in the function signature. Thus, better to add a check like the following before anything else:

```python
        if not all([value, item]):
            return None
```

RML-IAEA replied:
Thank you, this is now included.

xispa committed:
Flake8

xispa committed:
Flake8

xispa committed:
Undo not related changes

xispa committed:
Adhere to new rule for i18n message declarations

xispa committed:
Cleanup

xispa reviewed (changes requested):
Thanks, you are almost there!. Just two minor changes (see comments) and please add an entry at CHANGES.rst. Also, **do a pull first**, cause I've merged 2.x into your branch and done some commits to make Flake8 happy.

RML-IAEA committed:
Implemented xispa requested changes on Jan 19

Implemented xispa requested changes on Jan 19. see https://github.com/senaite/senaite.core/pull/2201

RML-IAEA commented:
Hi @xispa, 
Thank you for al your help. I have implemented all changes.

NiallMurphy-IAEA committed:
Merge branch 'senaite:2.x' into unitSelection

xispa committed:
Typo

xispa committed:
Remove whitespace line

xispa committed:
Do not translate whitespace

xispa committed:
Metadata column getUnitChoices in analysis catalog is not needed

xispa committed:
Move changelog entry to top

xispa committed:
Better description
